### ISSUE TITLE
Package/Download private JDK for jdt.ls

### ISSUE THREAD

**Issue by bagrupe** [redhat-developer/vscode-java#1682]
In a discussion within #1672 I suggested to package or download a 'private' JDK for jdt.ls.

This would have several benefits

- No need to explicitly let the user download a JDK distribution (many of which are zips without Installer and integration into the OS, which results in the need to manually set java.home) --> as a workaround the java-coding-pack works but not all users will realize that when they download vanilla VSCode and Install the Java extension pack directly from within VSCode
- The last version bump from 8 -> 11 was a source of frequent questions/issues and any such future bump will not be handled by the coding-pack for existing VSCode users
- vscode-java could move fast and always bundle the latest JDK with jdt.ls, so non-build-system-projects could always utlize the latest JDK, that's especially beneficial for new users which might expect the latest features to work automagically
- eventually the VSCode 'java.home' setting could be made obsolete as @Eskibear suggested in #1672 

Drawbacks would be
- When packaged the download size would explode
- When downloaded at first start/update you'll need an active internet connection

But I think the pros outweigh the cons in this case.

**Comment by 0dinD** [comment]
I agree that this would provide many benefits to the user experience, and it would put an end to the debate surrounding the JDK11 requirement. But if this is done, I am strongly opposed to packaging the JDK with the extension itself, for a number of reasons. Mainly, it would bloat the installation size by a large amount, especially since we would have to include a JDK for each OS, seeing as you can't provide different extension packages based on the OS. This is especially annoying if you've already got a suitable JDK installed - why should I be forced to download another one specifically for the extension? It's what IntelliJ does, but I think one of the strong selling points for this extension is how lightweight it is. 

With that said, I do think it would be nice to provide the **option** to automatically download a JDK for the extension which it can use for running the language server. Maybe there could be a notification dialog asking the user whether they want the extension to download a JDK, which only shows up if no suitable JDK is installed on the system? 

Also, I noticed that the Eclipse IDE now has an option to install a JDK, since it also needs JDK11 since the 2020-09 release. After looking closer at it, I found that the installer seems to be using [Eclipse JustJ](https://www.eclipse.org/justj/), which might be worth taking a look at when implementing this feature. It seems to be tailored toward running Eclipse applications and can produce smaller JDKs with `jlink` and the modular system introduced in JDK9. If this can significantly reduce the JDK size then ***maybe*** it would be OK to package it with the extension, but I would still prefer if it was an optional download.

**Comment by rgrunber** [comment]
The smallest possible JustJ JRE is 16-18MB (assuming jre.base.stripped is enough, https://download.eclipse.org/justj/jres/15/downloads/latest/ ) depending on the platform. It's nice to have the option but it would still be quite large to package >= 3 of these.

If the JDK is download (automatically, or via user selection), then size becomes less of a concern, though maybe we'll stick to JustJ in the automated case.

I've seen some vscode plugin that shows a nice UI for selecting and downloading particular java runtimes, so if we could use something like that, or some underlying common library that would be great.

**Comment by 0dinD** [comment]
@rgrunber I agree, it would be much better to download a JDK on-demand.

> I've seen some vscode plugin that shows a nice UI for selecting and downloading particular java runtimes, so if we could use something like that, or some underlying common library that would be great.

I think you're referring to the [Configure Java Runtime view](https://code.visualstudio.com/docs/java/java-tutorial#_settings-for-the-jdk) from [microsoft/vscode-java-pack](https://github.com/microsoft/vscode-java-pack)? It already has some nice features, but the install process is still manual, and I don't think the extension attempts to detect whether the user needs to download a JDK for jdt.ls. Maybe you can communicate with the Java extension pack developers about this issue? In a way I think it makes more sense to provide the feature via this extension, since not all developers will install the Java pack. Then again, the feature is mostly intended for newcomers, who will likely install the Java extension pack anyways.
